exercises/sat/descr.md

```diff
@@ -0,0 +1,616 @@
+
+The function `decompose` is expected to convert `s.f \/ c` to CNF, via
+suitable calls to `new_var` and `new_clause`. It returns no result.
+
+The function `clause` is expected to convert `s.f \/ c` to a clause,
+which it returns. It is also allowed to call `new_var` and `new_clause`
+so as to construct other clauses on the side.
+
+*Hint.* The easy case in `decompose` is when `s.f` is a conjunction, because
+it then suffices to handle each conjunct independently. The easy case in
+`clause` is when `s.f` is a disjunction, because it then suffices to deal with
+each disjunct in turn and combine the results into a single clause.
+
+## Recreation: Intersection of Two Sorted Lists
+
+**Question 5.** Implement a function `intersect`
+of type `int list -> int list -> bool`
+such that `intersect xs ys` is `true`
+if and only if the two *sorted* lists
+`xs` and `ys` have a common element.
+
+## Representing Clauses
+
+We would now like to choose a representation of clauses in memory and
+implement the functions `empty`, `cons`, `new_clause` and `new_var`
+that the functor `CNF` (above) requires.
+
+We adopt the convention that a literal is represented as a pair of a Boolean
+polarity and a variable:
+
+```
+  type literal = bool * var
+```
+
+We adopt the convention that a clause is represented as a list of literals:
+
+```
+  type clause = literal list
+```
+
+**Question A.** (Ungraded.) Implement the constant `empty` and the constructor
+function `cons`.
+
+We adopt the convention that clauses are numbered
+and are stored in an infinite array `clauses`.
+Infinite arrays are described at the very end of this exercise.
+
+**Question B.** (Ungraded.) Implement the function `new_clause`, which records a
+new clause, and the function `count_clauses`, which indicates how many clauses
+have been recorded so far. Note that a clause that contains both the literal
+`x` and the opposite literal `~x` is unsatisfiable. In such a situation, the
+function `new_clause` must raise the exception `UNSAT`. *Hint:* to detect this
+situation, use `intersect`.
+
+**Question C.** (Ungraded.) Implement the function `new_var`, which creates and
+returns a new variable.
+
+## The Undo Trail
+
+A SAT solver searches a tree of candidate solutions to a Boolean constraint.
+Repeatedly, it picks a variable `x` whose value has not yet been decided, sets
+this variable to `true` (for instance), and explores the consequences of this
+decision, which form a subtree. If it appears after a while that this decision
+was wrong (that is, there is no solution in this subtree) then the algorithm
+must *backtrack*, that is, come back to the point where this decision was
+made, undo this decision, make the opposite decision (that is, set `x` to
+`false`), and explore the consequences of this new decision, which form
+another subtree.
+
+A solver usually maintains several mutable data structures, such as a set of
+undecided variables, a set of unit clauses, etc. When the solver backtracks to
```

Comment:
Did you define `unit clauses` previously?

Reply:
No. I have added a forward pointer within parentheses.